### ISSUE TITLE
Don't link with libc or libdl

### ISSUE THREAD

**Issue by afjoseph** [android-ndk/ndk#1032]
I'm using NDK r18 with clang.

I'd like to build a shared library that does not use an STL as [described here](https://developer.android.com/ndk/guides/cpp-support).

I am using CMake and my CMakeLists.txt flags look like this:
```
project(libhello VERSION 0.1 LANGUAGES C)

add_library(
    ${PROJECT_NAME}
    SHARED
    src/main/c/main.c
    )

target_compile_options(
    ${PROJECT_NAME}
    PUBLIC "-Wall" "-Wextra" "-pedantic" "-Werror" "-Wshadow" "-nostdlib"
    )

set(ANDROID_STL "none")
```

Based on what the link I posted mentions, `there should be no linking or licensing requirements` if I used `none` STL. I was wondering how to build a native library and completely drop linking for `libc` and `libdl` from that library, which is what `-nostdlib` should have done (if I am not mistaken)

To be clearer, here's a couple of outputs from readelf to prove what I mentioned

```
[0] % readelf -a mylib.so

Dynamic section at offset 0xee8 contains 27 entries:
  Tag        Type                         Name/Value
 0x00000001 (NEEDED)                     Shared library: [libdl.so]
 0x00000001 (NEEDED)                     Shared library: [libc.so]
...
...
...
Symbol table '.dynsym' contains 8 entries:
   Num:    Value  Size Type    Bind   Vis      Ndx Name
     0: 00000000     0 NOTYPE  LOCAL  DEFAULT  UND
     1: 00000000     0 FUNC    GLOBAL DEFAULT  UND __cxa_atexit@LIBC (2)
     2: 00000000     0 FUNC    GLOBAL DEFAULT  UND __cxa_finalize@LIBC (2)
     3: 00000000     0 FUNC    GLOBAL DEFAULT  UND __stack_chk_fail@LIBC (2)
     4: 000004d0   154 FUNC    GLOBAL DEFAULT   12 Java_com_whatever
     5: 00002004     0 NOTYPE  GLOBAL DEFAULT  ABS _edata
     6: 00002004     0 NOTYPE  GLOBAL DEFAULT  ABS _end
     7: 00002004     0 NOTYPE  GLOBAL DEFAULT  ABS __bss_start
```

**Comment by enh** [comment]
yeah, when we said "there should be no linking or licensing requirements" we meant "caused by an STL". you still need libc for the stack protector (`__stack_chk_fail`) and support for running ELF destructors (`__cxa_atexit`/`__cxa_finalize`). libdl's admittedly a bit bogus, though i'm surprised we don't also have a `DT_NEEDED` for libm --- the OS build system always gives you libc+libm+libdl just because it's easier than making people remember to add -lm if they use `<math.h>` or `-ldl` if they use `<dlfcn.h>`.

(if we could start again from scratch, i'd put all the stuff from libm and libdl inside libc, and not have separate libraries at all, just like Android already doesn't have libpthread or librt. but it's way too late for that now.)

anyway, yeah, back to your question... would you be happy if we changed the docs to say something like:
```
There is also the option to have no STL. There are no STL linking or licensing
requirements in that case. No C++ standard headers are available.
```
?

**Comment by afjoseph** [comment]
Hey @enh. I'm completely fine with the wording. I realize the need for `__cxa_atexit` and the rest. I was a just a bit confused about the forced inclusion of libm and libdl (I only got rid of `libm.so` by forcefully doing `set(CMAKE_C_STANDARD_LIBRARIES "")`)

This clears things up. To conclude, `libc` is necessary for C bootstrapping and teardown and the rest is there to prevent folks from shooting themselves in the foot 👍 

Closing this. Much obliged!
